Fix validation of non-null Interface member

corydolphin opened the pull request:
If an interface has a member which is a non-null interface, an implementation of that interface with a concrete implementation of that interface fails.

I.E.

```GraphQL
interface TestMemberInterface {
  field: String
 }

type TestMemberInterfaceImpl implements TestMemberInterface {
  field: String
 }

interface TestInterface {
  field: TestMemberInterface!
}

type TestInterfaceImpl implements TestInterface {
  field: TestMemberInterfaceImpl!
}
```

Will fail to validate with error
```
object type 'TestInterfaceImpl' does not implement interface 'TestInterface' because field 'field' is defined as 'TestMemberInterfaceImpl!' type and not as 'TestMemberInterface!' type
```

bbakerman commented on src/main/java/graphql/schema/validation/ObjectsImplementInterfaces.java:
```diff
@@ -1,14 +1,6 @@
-import graphql.schema.GraphQLOutputType;
-import graphql.schema.GraphQLType;
-import graphql.schema.GraphQLUnionType;
+import graphql.schema.*;
```
Can you please fix up * imports to be how they were before

corydolphin replied:
Done! Apologies for the churn, my editor automatically updated the imports.

bbakerman commented:
I think we want this change.  However I am still a little confused because the examples are not 100% clear.

The reference implementation has :

https://github.com/graphql/graphql-js/blob/master/src/type/schema.js#L286

https://github.com/graphql/graphql-js/blob/master/src/utilities/typeComparators.js#L57

Can we put valid examples in place to ensure we are encountering what we should because the invalid examples are throwing me

corydolphin commented on src/test/groovy/graphql/schema/validation/ObjectsImplementInterfacesTest.groovy:
```diff
@@ -283,4 +283,38 @@ class ObjectsImplementInterfacesTest extends Specification {
+    def "field is a non null object"() {
+        given:
+        GraphQLInterfaceType memberInterface = newInterface()
+                .name("TestMemberInterface")
```
Good catch. I was trying to reduce noise, sorry for the confusion.

andimarek replied:
sorry, but not sure I get this: this interface has one field, so what is the problem?

corydolphin replied:
Bbakerman's comment was in reference to a previous revision of the change. 

The "Problem" is as stated in the description. The schema I provided there, and the test case here fails without the associated change to `ObjectsImplementInterfaces`

corydolphin commented:
Updated:

1. I separated the renaming to a separate PR https://github.com/graphql-java/graphql-java/pull/748
2. I have updated the description's schema to match the test's.

As for the content of the change, I actually think that the way https://github.com/graphql/graphql-js/blob/master/src/utilities/typeComparators.js#L57 implements it is clearer (and was my initial though), but I tried to match the fall-through negative case style. Logically to me, nullability is one of the highest order bits, so it should logically be the first check.
